jetty-websocket/websocket-jetty-tests/src/test/java/org/eclipse/jetty/websocket/tests/server/CloseInOnCloseEndpointNewThread.java

```diff
@@ -0,0 +1,31 @@
+//
+// ========================================================================
+// Copyright (c) 1995-2020 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under
+// the terms of the Eclipse Public License 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0
+//
+// This Source Code may also be made available under the following
+// Secondary Licenses when the conditions for such availability set
+// forth in the Eclipse Public License, v. 2.0 are satisfied:
+// the Apache License v2.0 which is available at
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.websocket.tests.server;
+
+import org.eclipse.jetty.websocket.api.StatusCode;
+
+public class CloseInOnCloseEndpointNewThread extends AbstractCloseEndpoint
+{
+    @Override
+    public void onWebSocketClose(int statusCode, String reason)
+    {
+        new Thread(() -> getSession().close(StatusCode.SERVER_ERROR, "this should be a noop")).start();
+        super.onWebSocketClose(statusCode, reason);
```

Comment:
This test should block and wait for the thread (and therefore the `close()`) to finish, and then return from `onWebSocketClose()`. In this way we guarantee that we don't deadlock.

Reply:
Why expected to be different to using `CloseInOnCloseEndpoint`, which still does the blocking close but is done in the same thread?

Reply:
Because you are stating in the test what is the behavior that must happen.

If you revert the code to use `SharedBlockingCallback` or a `Lock`, this test will fail because it waits indefinitely.

Your comments assumes how the code works and so you ask "why writing a test since I know how the code works".

That is exactly why you write tests.

Reply:
I'm not asking "why writing a test since I know how the code works", I already had the test to reproduce the deadlock, which is the same thing as this without the 
 `new Thread(() -> ` part. Was just asking why we want to start a new thread to do the close in this version of it.

Reply:
Because testing it with 2 threads will ensure that concurrent closes will work, while testing from a single thread does not - locks employed could be reentrant.